src/client/Client.cc

```diff
@@ -5534,12 +5534,11 @@ void Client::handle_cap_trunc(MetaSession *session, Inode *in, const MConstRef<M
 
   uint64_t size = m->get_size();
   if (in->is_fscrypt_enabled()) {
-    size = std::stoll(std::string(std::rbegin(m->fscrypt_file),
-                                  std::rend(m->fscrypt_file)));
+    size = std::accumulate(m->fscrypt_file.rbegin(), m->fscrypt_file.rend(), 0ll,
```

Comment:
@chrisphoffman FYI

I'd like to clarify the semantics of the `ceph.fscrypt.file` vxattr vs its storage format in `fscrypt_file`
Other attributes tend to generate and consume their xattr values as strings which get parsed on set and generated on get. For example, see the layout pool:

```
size_t Client::_vxattrcb_layout_pool(Inode *in, char *val, size_t size)
{
  size_t r;
  objecter->with_osdmap([&](const OSDMap& o) {
      if (o.have_pg_pool(in->layout.pool_id))
	r = snprintf(val, size, "%s", o.get_pool_name(
		       in->layout.pool_id).c_str());
      else
	r = snprintf(val, size, "%" PRIu64, (uint64_t)in->layout.pool_id);
    });
  return r;
}
```

It does seem reasonable to keep the file size as an encoded uint64, though it would be less confusing to keep it as such a type explicitly. I'm not comfortable with the current change which assumes little-endian and performs the decoding manually (+1 for std::accumulate, though), especially outside a common fscrypt file size getter. I think I saw this field interpreted as uint64 in Chris's branch. I doubt we would like to perform vector of bytes conversion to an integer every time we need the value of the fscrypt file size.

On the side of the setter, however, how common is it that we pass a pointer to a platform integer for the xattr value? I don't see any validation in the setter.

```
int Client::_vxattrcb_fscrypt_file_set(Inode *in, const void *val, size_t size,
				       const UserPerm& perms)
{
  struct ceph_statx stx = { 0 };
  std::vector<uint8_t>	aux;

  aux.resize(size);
  memcpy(aux.data(), val, size);

  return _do_setattr(in, &stx, CEPH_SETATTR_FSCRYPT_FILE, perms, nullptr, &aux);
}
```

Finally, to confuse us even more, there is a code in the libcephfs test that sets `ceph.fscrypt.file` to a string `"foo"`

```
  ASSERT_EQ(0, ceph_fsetxattr(cmount, fd, "ceph.fscrypt.file", "foo", 3, CEPH_XATTR_CREATE));
```

Reply:
Yeah, why we made the `fscrypt_file` as a vector before just because this should be opaque for cephfs and the client not supporting fscrypt.

Once we are here to support fscrypt in libcephfs, I am fine to switch `fscrypt_file` to `u64`.

@vshankar @jtlayton What do you think ?

Reply:
It should be a platform integer just like all other internal inode parameters. The trick is to have it properly deserialized in the decode, and be backward compatible with current versions. Maybe this accumulate code will be handy there, or maybe we can make use of `ceph_le64`.

Reply:
> endianness matters here

@jtlayton you are right, I missed that this field can be set on the MDS via a setattr request. I thought this was all handled by the client, but that wasn't the case.

Do we have any convention for binary integral data sent to the MDS by the clients as byte payload? I suspect that if we had one it would be similar to what the encoder/decoder is doing, which is to send binary data as little endian over the wire.

Reply:
> I don't have an objection, but you do need to bear in mind that endianness matters here. If you're going to store it as a uint64_t, then will it be little-endian?

Yeah, correct.

Reply:
I will make this to be tested and merged via https://github.com/ceph/ceph/pull/54725, and will keep this PR to do a following improvement by switching the `fscrypt_file` to `uint64_t`.

Reply:
> Maybe this accumulate code will be handy there, or maybe we can make use of ceph_le64

I use that in wip-fscrypt branch here: https://github.com/chrisphoffman/ceph/tree/wip-fscrypt . 

for example:
`size = *(ceph_le64 *)st->fscrypt_file.data();` 

I have no problem with fixing it like @lxbsz proposes and then make the switch when my branch is ready.

Reply:
Agreed.